### PR TITLE
sci-astronomy/pal: Fix call to undeclared library function strlcpy

### PR DESCRIPTION
**Pull request by listout**
Closes: https://bugs.gentoo.org/898106

### CODE REVIEW

File: sci-astronomy/pal/pal-0.9.8-r1.ebuild [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+DESCRIPTION="Positional Astronomy Library"
+HOMEPAGE="https://github.com/Starlink/pal"
+SRC_URI="https://github.com/Starlink/${PN}/releases/download/v${PV}/${P}.tar.gz"
+
+LICENSE="GPL-2"
+SLOT="0"
+KEYWORDS="~amd64 ~x86 ~amd64-linux ~x86-linux"
+IUSE="static-libs"
+RDEPEND="sci-astronomy/erfa:="
+DEPEND="${RDEPEND}"
```
Comment (AndrewAmmerlaan) [inline sci-astronomy/pal/pal-0.9.8-r1.ebuild:15]: Shouldn't this depend on `dev-libs/libbsd` then? Since this is wat provides `bsd/string.h`.

Comment (listout) [reply sci-astronomy/pal/pal-0.9.8-r1.ebuild:15]: Even with dev-libs/libbsd I can repro due the error. From what it looks like, HAVE_BSD_STRING_H is not properly set by the configure script hence the inclusion fails/doesn't happen. I might be wrong.

Comment (AndrewAmmerlaan) [reply sci-astronomy/pal/pal-0.9.8-r1.ebuild:15]: Right, but if the header is not there the build will fail as well, so if we patch it like this we have to DEPEND on libbsd.

But we only need this when building with clang/musl right? This way we also change the build if we compile with gcc, probably what we want then instead is to set HAVE_BSD_STRING_H if the compiler is clang. Though then still we have to DEPEND on libbsd so maybe this is not worth the extra effort.

Comment (listout) [reply sci-astronomy/pal/pal-0.9.8-r1.ebuild]: Aha ... got it. How does it look now?

File: sci-astronomy/pal/files/pal-0.9.8-fix-strlcpy-musl.patch [hunk 2]
```diff
@@ -0,0 +1,13 @@
+Bug: https://bugs.gentoo.org/898106
+bsd/string.h is available on most libc's
+--- a/palDfltin.c
++++ b/palDfltin.c
+@@ -119,7 +119,7 @@ static int ISBLANK( int c ) {
+ 
+ #endif
+ 
+-#ifdef HAVE_BSD_STRING_H
```
Comment (thesamesam) [inline sci-astronomy/pal/files/pal-0.9.8-fix-strlcpy-musl.patch:9]: I don't see the need for this patch, unless the `HAVE_BSD_STRING_H` test is failing on musl even when it's installed. But then we should fix the configure test.

File: sci-astronomy/pal/files/pal-0.9.8-fix-strlcpy-musl.patch [hunk 3]
```diff
@@ -0,0 +1,13 @@
+Bug: https://bugs.gentoo.org/898106
+bsd/string.h is available on most libc's
```
Comment (thesamesam) [inline sci-astronomy/pal/files/pal-0.9.8-fix-strlcpy-musl.patch:2]: `string.h` should always exist (POSIX), but `bsd/string.h` is from libbsd.

The actual issue here is "does the libc's string.h provide strlcpy?"

The answer is:
* musl does under _GNU_SOURCE
* >=glibc-2.38 does under _GNU_SOURCE
* both of them will provide it in general "soon" as it's in the next edition of POSIX

**Review by thesamesam — CHANGES_REQUESTED** [review]

File: sci-astronomy/pal/files/pal-0.9.8-fix-strlcpy-musl.patch (hunk 2 above)
Comment (listout) [reply sci-astronomy/pal/files/pal-0.9.8-fix-strlcpy-musl.patch:9]: yes that seems to be the case Sam.

File: sci-astronomy/pal/files/pal-0.9.8-fix-strlcpy-musl.patch (hunk 3 above)
Comment (listout) [reply sci-astronomy/pal/files/pal-0.9.8-fix-strlcpy-musl.patch:2]: So should it be using something like
```c
#define _GNU_SOURCE
#include <string.h>
```
instead?

**Comment by AndrewAmmerlaan** [comment]
Gentle ping

**Comment by listout** [comment]
I have updated the patch a bit based on what Sam and Andrew suggested/said. Please have a look (requesting feedback).

**Review by AndrewAmmerlaan — APPROVED** [review]
Yes this new patch makes more sense to me.

@thesamesam do you agree?

File: sci-astronomy/pal/pal-0.9.8-r1.ebuild [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+SRC_URI="https://github.com/Starlink/${PN}/releases/download/v${PV}/${P}.tar.gz"
+
+LICENSE="GPL-2"
+SLOT="0"
+KEYWORDS="~amd64 ~x86 ~amd64-linux ~x86-linux"
+IUSE="static-libs"
+RDEPEND="sci-astronomy/erfa:="
+DEPEND="
+	${RDEPEND}
+	dev-libs/libbsd
```
Comment (AndrewAmmerlaan) [inline sci-astronomy/pal/pal-0.9.8-r1.ebuild:17]: Though now we don't have to depend on libbsd any more.

**Comment by AndrewAmmerlaan** [comment]
Actually, now that I test this, it still fails for me with the same error with clang-16 and 17.

`string.h` on my system (glibc-2.38-r7) looks like:
```
...
#ifdef __USE_MISC
/* Copy at most N - 1 characters from SRC to DEST.  */
extern size_t strlcpy (char *__restrict __dest,
const char *__restrict __src, size_t __n)
__THROW __nonnull ((1, 2)) __attr_access ((__write_only__, 1, 3));
...
```
If I `#define __USE_MISC` instead, then it does work. But this I have not tested with musl, we might need a different `#define` for glibc versus musl.

**Comment by listout** [comment]
I tested the PR against glibc clang profile, https://0x0.st/HwHG.log

Can specify the test env.

**Comment by AndrewAmmerlaan** [comment]
I've adjusted the patch to define both `_GNU_SOURCE` and `__USE_MISC`. This should ensure that it works in both cases.

**Pull request closed without merging**